Refactor test to cope with known docs counts error

craigtaverner opened the pull request:
Terms aggs can generate incorrect counts when combining results from multiple segments. Recent tests enhancements now tests multiple segments introducing flakiness for these cases. This fix adjusts the assertions to cope with that scenario.

craigtaverner committed:
Refactor test to cope with known docs counts error

Terms aggs can generate incorrect counts when combining results from
multiple segments. Recent tests enhancements now tests multiple
segments introducing flakiness for these cases. This fix adjusts the
assertions to cope with that scenario.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

not-napoleon commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java:
```diff
@@ -921,22 +922,28 @@ private <T> void termsAggregator(
-                            assertEquals(String.valueOf(expected.getKey()).substring(1), actual.getKey());
-                        } else {
-                            assertEquals(expected.getKey(), actual.getKey());
+                        long expectedCount = expectedCounts.getOrDefault(actual.getKey(), 0L);
```
We should definitely add a comment here.  This was confusing enough the first time through, I don't want to have to figure this out again next time I touch it.

not-napoleon commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java:
```diff
@@ -921,22 +922,28 @@ private <T> void termsAggregator(
+                        long expectedCount = expectedCounts.getOrDefault(actual.getKey(), 0L);
+                        if (aggTestConfig.splitLeavesIntoSeparateAggregators() == false || indexReader.leaves().size() == 1) {
+                            assertThat("Count mismatch for " + actual.getKey(), actual.getDocCount(), equalTo(expectedCount));
+                        } else if (order == false) {
```
I think it makes sense to force count ascending (aka `order == true`) into single segment mode.  Otherwise there's just no invariant we can test for it. I'm not 100% sure that testing count ascending in single segment mode proves anything, since prod rarely runs in single collector mode, but I also don't think we should allow sort by count ascending, so...

martijnvg replied:
When either `aggTestConfig.splitLeavesIntoSeparateAggregators() == false` or `indexReader.leaves().size() == 1` then we can assert docCounts when `order` is `false` or `true`?

craigtaverner replied:
Yes, no matter the sort order, if there is only one sub-aggregator, there are no counts errors, so we can assert on every value. The older version of this test made this assumption, and also only used a single sub-aggregator. The new test allows for multiple aggregators, but can only make the assertions when the error is unbounded.

craigtaverner replied:
So I've made a commit which does two things, adds comments clarifying all this, as well as implements the suggestion to force single-aggregator when order == true (ie. allow multiple aggregators for order == false).

not-napoleon reviewed (commented):
We still need to replicate this behavior down into the multivalued section of the test, but I think this is an improvement.  It lets us test the multi-collector path, which is much more likely to be what we hit in production, and also puts some checking testing on `docCountError`, which is nice.  I co-authored this, so I'm not going to approve it, but leaving notes for the future.

martijnvg commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java:
```diff
@@ -921,22 +922,28 @@ private <T> void termsAggregator(
+                            assertThat("Count mismatch for " + actual.getKey(), actual.getDocCount(), equalTo(expectedCount));
+                        } else if (order == false) {
+                            long diff = Math.abs(expectedCount - actual.getDocCount());
+                            assertThat("Count error too large for " + actual.getKey(), diff, lessThanOrEqualTo(docCountError));
```
When I look at this if / else if clauses it seems there are cases where nothing is asserted.
Can the count error always be asserted? In cases with a single segment or when splitLeavesIntoSeperateAggregators is disabled then count error should be zero? 

If so then we only conditionally assert the docCount against expectedDocCount if the test conditions allow for this.

craigtaverner committed:
Avoid order=true and multiple aggregators

This combination leads to unbounded counts errors, so no per-record
assertions can be made. We decided to skip this case for now.

craigtaverner commented:
> We still need to replicate this behavior down into the multivalued section of the test, but I think this is an improvement. It lets us test the multi-collector path, which is much more likely to be what we hit in production, and also puts some checking testing on `docCountError`, which is nice. I co-authored this, so I'm not going to approve it, but leaving notes for the future.

I'd love a failing test to track before making a change to the multi-valued section. I'll try find one.

craigtaverner commented:
Bad news. When running with `-Dtests.iters=1000` I had zero failures from the multi-valued part of the test, but 9 failures on the new maximum error check. It seems as if the maximum error is not reliably bounded by the `docCountError`.

not-napoleon committed:
Comments and extra data on failure

not-napoleon committed:
fix expected value range

not-napoleon committed:
Remove seed

craigtaverner commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java:
```diff
@@ -938,8 +939,7 @@ private <T> void termsAggregator(
                     HashMap<Object, Long> expectedCounts = new HashMap<>();
-                    for (int i = 0; i < size; i++) {
-                        Map.Entry<T, Integer> expected = expectedBuckets.get(i);
+                    for (Map.Entry<T, Integer> expected : expectedBuckets) {
```
Ah, because of possible order changes we need to use the entire expected set to compare against!

craigtaverner commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -590,6 +590,7 @@ private <A extends InternalAggregation, C extends Aggregator> A searchAndReduce(
         try {
             if (randomBoolean() && aggs.size() > 1) {
                 // sometimes do an incremental reduce
+                logger.info("Doing incremental reduce");
```
Do we need this line? Perhaps it was there for temporary debugging?

martijnvg replied:
This was added in this PR. Maybe to debug something. I think we should not add it?

craigtaverner replied:
I removed it.

craigtaverner committed:
Fixed formatting errors

and simplified error message and comment

martijnvg commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java:
```diff
@@ -921,22 +923,50 @@ private <T> void termsAggregator(
-                            assertEquals(String.valueOf(expected.getKey()).substring(1), actual.getKey());
+                        long bucketDocCountError = actual.getDocCountError();
+                        assertThat(
+                            "Bucket doc count error is greater than the maximum doc count",
```
This assertion tripped in ci. Maybe this can't always be asserted?

craigtaverner replied:
When I run with `-Dtests.iters=1000` I get about 10 failures on this, so about 1% of the time. So this does seem to be an issue.

craigtaverner committed:
Removed logging line

craigtaverner commented:
Closing this PR because the main concern with flaky tests was resolved with another PR, and this one focused only on extending test coverage to multiple aggregators. Due to remaining flaky tests with that additional coverage, we postpone working on this for now. The new flaky test failures do hint at a possible bug in the docCountError calculation, so we would like to revisit this at some point.
